### PR TITLE
Improves Type Annotations

### PR DESCRIPTION
**Pull request by WaelKarkoub**
## What this does
I want to contribute more often so I thought maybe a PR that goes through "obvious" missing annotations could be a good first PR.

## How it was tested
- ran `poetry run pytest tests/`

- - -
This change is [Reviewable](https://reviewable.io/reviews/huggingface/lerobot/252)

### CODE REVIEW

**Comment by WaelKarkoub** [comment]
Hi @alexander-soare, for some reason I'm not able to tag anyone to review my PR, let me know what I should do

**Comment by alexander-soare** [comment]
Thanks for the PR! I tagged myself and will take a look early this week.

File: lerobot/scripts/eval.py [hunk 1]
```diff
@@ -342,11 +350,13 @@ def render_frame(env: gym.vector.VectorEnv):
         # Maybe render video for visualization.
         if max_episodes_rendered > 0 and len(ep_frames) > 0:
             batch_stacked_frames = np.stack(ep_frames, axis=1)  # (b, t, *)
+            assert isinstance(videos_dir, Path)
```
Comment (alexander-soare) [inline lerobot/scripts/eval.py:353]: Perhaps we should check this at the very start, and raise a `ValueError` if the caller tries to provide `max_episodes_rendered > 0` without `videos_dir`. What do you think?

File: lerobot/scripts/eval.py [hunk 2]
```diff
@@ -535,6 +545,8 @@ def main(
     else:
         # Note: We need the dataset stats to pass to the policy's normalization modules.
         policy = make_policy(hydra_cfg=hydra_cfg, dataset_stats=make_dataset(hydra_cfg).stats)
+
+    assert isinstance(policy, nn.Module)
```
Comment (alexander-soare) [inline lerobot/scripts/eval.py:549]: I have 2 comments earlier about this check. Leaving this note here to make sure we resolve.

File: lerobot/scripts/eval.py [hunk 3]
```diff
@@ -623,7 +635,7 @@ def main(
             )
 
         main(
-            pretrained_policy_path=pretrained_policy_path,
+            pretrained_policy_path=str(pretrained_policy_path),
```
Comment (alexander-soare) [inline lerobot/scripts/eval.py:638]: Actually, it looks like we should type hint this as a `Path` because we use the `__div__` operator for `Path` in the logic.

Btw, have you tested this? I assume it should raise an exception.

File: lerobot/scripts/push_dataset_to_hub.py [hunk 4]
```diff
@@ -209,6 +212,7 @@ def push_dataset_to_hub(
         save_meta_data(info, stats, episode_data_index, meta_data_dir)
 
     if not dry_run:
+        # FIXME: token needs to be a str | None
```
Comment (alexander-soare) [inline lerobot/scripts/push_dataset_to_hub.py:215]: cc @Cadene It seems `token` should be a `str | None` according toe the `Dataset.push_to_hub` API. Could this be a problem? How should @WaelKarkoub address this?

File: poetry.lock [hunk 5]
```diff
@@ -1,4 +1,4 @@
-# This file is automatically @generated by Poetry 1.8.1 and should not be changed by hand.
```
Comment (alexander-soare) [inline poetry.lock:1]: @aliberts There are a lot of changes here without changes to `pyproject.toml`. Should we expect this?

File: lerobot/scripts/train.py [hunk 6]
```diff
@@ -302,7 +304,7 @@ def train(cfg: DictConfig, out_dir: str | None = None, job_name: str | None = No
         dataset_stats=offline_dataset.stats if not cfg.resume else None,
         pretrained_policy_name_or_path=str(logger.last_pretrained_model_dir) if cfg.resume else None,
     )
-
+    assert isinstance(policy, nn.Module)
```
Comment (alexander-soare) [inline lerobot/scripts/train.py:307]: I have comments earlier about this assertion. Leaving this here to block till resolved.

**Review by alexander-soare — CHANGES_REQUESTED** [review]
Thanks so much! I left some comments.

File: lerobot/scripts/push_dataset_to_hub.py (hunk 4 above)
Comment (Cadene) [reply lerobot/scripts/push_dataset_to_hub.py:215]: Nice finding. I will take care of it in another PR.

Comment (alexander-soare) [reply lerobot/scripts/push_dataset_to_hub.py:215]: Thanks @Cadene. @WaelKarkoub in that case could you please change to
```suggestion
        # TODO(rcadene): token needs to be a str | None
```

File: poetry.lock (hunk 5 above)
Comment (aliberts) [reply poetry.lock:1]: There shouldn't be any changes to `poetry.lock`, it's probably due to having run something like `poetry lock`

I'd suggest to revert changes on the lock file.

I'm actually thinking of removing the `poetry.lock` file entirely from the repo.

File: lerobot/common/policies/policy_protocol.py [hunk 7]
```diff
@@ -49,20 +49,23 @@ def reset(self):
 
         Does things like clearing caches.
         """
+        ...
```
[the remark this answers is not in the document]
Comment (WaelKarkoub) [reply lerobot/common/policies/policy_protocol.py:52]: I'm following PEP 544 for protocol definition/style

- https://peps.python.org/pep-0544/#defining-a-protocol
- https://typing.readthedocs.io/en/latest/spec/protocol.html#protocols

File: lerobot/scripts/eval.py [hunk 8]
```diff
@@ -105,7 +105,7 @@ def rollout(
 
     Args:
         env: The batch of environments.
-        policy: The policy.
+        policy: The policy. Must be a PyTorch nn module.
```
[the remark this answers is not in the document]
Comment (WaelKarkoub) [reply lerobot/scripts/eval.py:108]: I'm on the same boat as you, the proper fix would be allowing `Policy` to inherit `nn.Module` (i.e. `class Policy(nn.Module, Protocol)`). 

Correct me if wrong `Policy | nn.Module` pretty much means `Union[Policy, nn.Module]`. But looking at the implementation, we expect `policy` to be `Policy` _and_ `nn.Module`, hence the requirement for inheritance. (This is where I wished for python to have a trait system like the one rust has :) )

File: lerobot/scripts/eval.py [hunk 9]
```diff
@@ -116,6 +116,7 @@ def rollout(
     Returns:
         The dictionary described above.
     """
+    assert isinstance(policy, nn.Module), "Policy must be a PyTorch nn module."
```
[the remark this answers is not in the document]
Comment (WaelKarkoub) [reply lerobot/scripts/eval.py:119]: Due to the way `Policy` was defined, we can't guarantee that anyone who implements the `Policy` protocol will also implement/inherit `nn.Module`. My goal for this PR is to make sure if we fail due wrong types we fail early, but there is a better way to solving this problem which I thought it needs to be done in a separate PR (`Policy` class inheriting `nn.module`)

File: lerobot/scripts/eval.py [hunk 10]
```diff
@@ -194,6 +195,7 @@ def rollout(
         "reward": torch.stack(all_rewards, dim=1),
         "success": torch.stack(all_successes, dim=1),
         "done": torch.stack(all_dones, dim=1),
+        "observation": None,
```
[the remark this answers is not in the document]
Comment (WaelKarkoub) [reply lerobot/scripts/eval.py:198]: it lowers the cognitive load by knowing exactly what could be returned by this rollout function, instead of having to hunt it down by reading the entire function first. That's why I always favor "strongly" typed return objects (e.g. dataclasses, pydantic models, protobuf, etc...).

But this is a preference as you said, more than happy to revert it.

File: lerobot/scripts/eval.py (hunk 3 above)
Comment (WaelKarkoub) [reply lerobot/scripts/eval.py]: Good catch, I haven't tested this but `pretrained_policy_path` had `str | None` type. Fixed in the next commit

File: lerobot/scripts/eval.py [hunk 11]
```diff
@@ -285,7 +289,10 @@ def render_frame(env: gym.vector.VectorEnv):
         # this won't be included).
         n_steps = rollout_data["done"].shape[1]
         # Note: this relies on a property of argmax: that it returns the first occurrence as a tiebreaker.
-        done_indices = torch.argmax(rollout_data["done"].to(int), axis=1)  # (batch_size, rollout_steps)
+        done_indices = torch.argmax(
```
[the remark this answers is not in the document]
Comment (WaelKarkoub) [reply lerobot/scripts/eval.py]: I was able to test this, and they are indeed equivalent. I switched it to `dim` since it's documented to avoid further confusion.

File: lerobot/scripts/eval.py [hunk 12]
```diff
@@ -298,6 +305,7 @@ def render_frame(env: gym.vector.VectorEnv):
         all_successes.extend(batch_successes.tolist())
         all_seeds.extend(seeds)
 
+        # FIXME: episode_data is either None or it doesn't exist
```
[the remark this answers is not in the document]
Comment (WaelKarkoub) [reply lerobot/scripts/eval.py:313]: That is correct, it could be a separate PR. Trying to control the size of this PR

File: lerobot/common/policies/policy_protocol.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (alexander-soare) [reply lerobot/common/policies/policy_protocol.py:52]: Hmm. I think we should make this change if we decide we want to follow this style pointer. But we haven't made that determination, and on a higher level we currently want to avoid being that specific about style.

For instance, if this were amidst a bunch of other changes (like a whole new class) I'd probs leave it as is without comment. But here, we'd be changing the status quo without any overarching reason to do so. Mind if we leave it out?

File: lerobot/scripts/eval.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (alexander-soare) [reply lerobot/scripts/eval.py:119]: Hmm, okay, I mean I'm not 100% convinced because I'd say there are other things on a similar level that we are lax about checking. But I'm okay with this. To be clear though, if someone contributed some other function/class that uses a policy and _didn't_ put this check in, I wouldn't raise it in a review.

File: lerobot/scripts/eval.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (alexander-soare) [reply lerobot/scripts/eval.py:108]: I have one comment addressing this whole policy typing thing. Refer to that one for all of these threads please.

**Review by alexander-soare — CHANGES_REQUESTED** [review]
Thanks for revising! Still blocking based on @aliberts' ask.

File: poetry.lock (hunk 5 above)
Comment (WaelKarkoub) [reply poetry.lock]: If I remember correctly, I updated the packages which could explain why the lock file changed. I reverted that change

**Review by alexander-soare — APPROVED** [review]
